### PR TITLE
REF/ENH: Console rework

### LINKED ISSUE #195: Race condition adding devices to TyphonConsole
Via the `typhon` logger:

```
[2019-03-19 22:28:59] - INFO -  Storing device Slits mfx_dg1_slits (prefix=MFX:DG1:JAWS, z=1019.3) ...
[2019-03-19 22:28:59] - INFO -  Adding / Modifying information for MFX:DG1:JAWS ...
[2019-03-19 22:28:59] - INFO -  Device Slits mfx_dg1_slits (prefix=MFX:DG1:JAWS, z=1019.3) has been succesfully added to the database
[2019-03-19 22:28:59] - INFO -  Loading 'mfx_dg1_valve_1' ...
[2019-03-19 22:28:59] - ERROR -  Unable to add device 'mfx_dg1_valve_1' to TyphonConsole.
Traceback (most recent call last):
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/client.py", line 59, in __init__
    self.backend = backend(**kwargs)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/backends/json_db.py", line 42, in __init__
    self.initialize()
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/backends/json_db.py", line 69, in initialize
    "a new database.".format(self.path))
PermissionError: File /tmp/e19cb4cf90f5de4bb26bb684d053998b already exists. Can not initialize a new database.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/typhon/tools/console.py", line 90, in add_device
    client = happi.Client(path=name, initialize=True)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/client.py", line 62, in __init__
    "a {} backend".format(backend)) from exc
happi.errors.DatabaseError: Failed to instantiate a <class 'happi.backends.json_db.JSONBackend'> backend
```

Inside the console:
```
ileNotFoundError                         Traceback (most recent call last)
<ipython-input-1-5ad3102e6c36> in <module>
----> 1 import happi; from happi.loader import from_container; client = happi.Client(path="/tmp/e19cb4cf90f5de4bb26bb684d053998b"); md = client.find_device(name="mfx_dg1_slits"); mfx_dg1_slits = from_container(md)

/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/client.py in find_device(self, **post)
    188         """
    189         logger.debug("Gathering information about the device ...")
--> 190         doc = self.find_document(**post)
    191         # Instantiate Device
    192         logger.debug("Instantiating device based on found information ...")

/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/client.py in find_document(self, **kwargs)
     90         """
     91         # Request information from backend
---> 92         post = self.backend.find(multiples=False, **kwargs)
     93         # Check result, if not found let the user know
     94         if not post:

/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/backends/json_db.py in find(self, _id, multiples, **kwargs)
    120         """
    121         # Load database
--> 122         db = self.load()
    123 
    124         # Search by _id, separated for speed

/reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0/lib/python3.6/site-packages/happi/backends/json_db.py in load(self)
     76         Load the JSON database
     77         """
---> 78         with open(self.path, 'r') as f:
     79             return json.load(f)
     80 

FileNotFoundError: [Errno 2] No such file or directory: '/tmp/e19cb4cf90f5de4bb26bb684d053998b'
```

### LINKED ISSUE #243: Tools when not from Happi
## Expected Behavior
Make sure that tools such as Console, Striptool and etc work properly when a device was not created via Happi.

## Current Behavior

Console give an error about missing metadata.

### PR DESCRIPTION
**Pull request by klauer**
* Allow for 3 classes of devices to work in the console, and has simple tests for each:
   - Importable devices (importable by introspection of class name + `device._repr_info` is set appropriately) (*)
   - Fake devices
   - Happi client-accessible devices
* Closes #243 as best we can
* Closes #195 due to refactoring
* Separates out 'device creation' code into utils
* Attempts to clean the console code a bit and add some convenience methods

(*) This isn't always possible, you might get `__main__` as a module name or `tests` in the case of when running pytest, for example.

### CODE REVIEW

**Comment by klauer** [comment]
Tests failing due to happi not being released, so I just tagged a new version: https://github.com/pcdshub/happi/releases/tag/v1.5.0

File: typhos/tools/console.py [hunk 1]
```diff
@@ -61,62 +74,52 @@ def sizeHint(self):
...
+            self.jupyter_widget.kernel_manager.shutdown_kernel()
+            self.kernel_shut_down.emit()
+
+        client.stop_channels()
+        QtCore.QTimer.singleShot(0, cleanup)
+
+    def add_device(self, device):
+        # Add the device after a short delay to allow the console widget time
+        # to get initialized
+        self.show()
```
Comment (klauer) [inline typhos/tools/console.py:95]:
```suggestion
```

File: typhos/tools/console.py [hunk 2]
```diff
@@ -61,62 +74,51 @@ def sizeHint(self):
...
+
+        client.stop_channels()
+        QtCore.QTimer.singleShot(0, cleanup)
+
+    def add_device(self, device):
+        # Add the device after a short delay to allow the console widget time
+        # to get initialized
+        QtCore.QTimer.singleShot(
+            100, lambda device=device: self._add_device(device)
+        )
```
Comment (ZLLentz) [inline typhos/tools/console.py:97]: Timers with magic numbers always make me nervous

Comment (klauer) [reply typhos/tools/console.py:97]: Not sure what we could do better. Perhaps watch the IPython console output until we see a prompt `In [1]:` and emit a signal?

I dug a bit, but didn't see much else in the interface to aid in synchronization. Rather, programmatic interaction with the kernel in general seemed to be somewhat lacking.

Comment (ZLLentz) [reply typhos/tools/console.py:97]: If there's nothing better, there's nothing better. 🤷

**Comment by klauer** [comment]
I think it will always be a pain-point. While there is an alternative interface that allows for an in-process kernel, things like `time.sleep` inside that terminal block the entire process.

File: typhos/tools/console.py (hunk 2 above)
Comment (klauer) [reply typhos/tools/console.py:97]: Let me give the above a try and see how it works...

**Comment by klauer** [comment]
Poking around a bit with the initialization routine, the synchronization is tricky because it takes even longer than you'd expect to startup a kernel:

My debug messages: `It took 0.6457221508026123 secs to get the console ready` show approximately this value pretty consistently

Here's why I like this solution, instead of the dumb timer @ZLLentz rightfully pointed out as scary:
* You're guaranteed to see the `In [1]:` prompt at the beginning
* You'll get to see the code it takes to create your device

Why I don't like it:
* It's a lot of additional complexity

**Comment by hhslepicka** [comment]
So, I did the stupid mistake of typing `exit` at the console... 
After that it restarted the kernel but my device was no longer there. Is there something we can do to protect against this or re-add the devices to the console once the kernel restarts?

**Comment by klauer** [comment]
Good question, @hhslepicka - not that I see. If we could figure out how to make an ipython_profile sort of configuration, it could work. I haven't been able to find that hook, but would gladly accept help.

Failing that, we can add a button that says "restart kernel" and redoes the init code we do now.

**Comment by hhslepicka** [comment]
Here is a nice test that I runned...
Using `Bluesky` inside of the Typhos console scanning my two motors... :neckbeard: 
Screen Shot 2020-04-14 at 3 44 56 PM

**Comment by hhslepicka** [comment]
I was able to find a `custom_restart` flag... I am investigating it here... if you would like to go ahead and merge that in I can submit a PR later... otherwise I could take until tomorrow if you are willing to wait and if I don't get back to you, merge at will...

**Comment by klauer** [comment]
Nice find, @hhslepicka - thinking I might have missed something that indicated "kernel ready", I looked for other signals:

```
$ grep -RH -e '=\s*QtCore.Signal' .
./frontend_widget.py:    custom_interrupt_requested = QtCore.Signal()
./frontend_widget.py:    custom_restart_kernel_died = QtCore.Signal(float)
./frontend_widget.py:    custom_restart_requested = QtCore.Signal()
./frontend_widget.py:    executing = QtCore.Signal(object)
./frontend_widget.py:    executed = QtCore.Signal(object)
./frontend_widget.py:    exit_requested = QtCore.Signal(object)
./kernel_mixins.py:    kernel_restarted = QtCore.Signal()
./kernel_mixins.py:    started_channels = QtCore.Signal()
./kernel_mixins.py:    stopped_channels = QtCore.Signal()
./comms.py:    sig_is_closing = QtCore.Signal(object)
./client.py:    kernel_died = QtCore.Signal(object)
./client.py:    message_received = QtCore.Signal(object)
./jupyter_widget.py:    custom_edit_requested = QtCore.Signal(object, object)
./inprocess.py:    started = QtCore.Signal()
./inprocess.py:    stopped = QtCore.Signal()
./inprocess.py:    message_received = QtCore.Signal(object)
./inprocess.py:    kernel_died = QtCore.Signal()
./console_widget.py:    copy_available = QtCore.Signal(bool)
./console_widget.py:    redo_available = QtCore.Signal(bool)
./console_widget.py:    undo_available = QtCore.Signal(bool)
./console_widget.py:    custom_page_requested = QtCore.Signal(object)
./console_widget.py:    font_changed = QtCore.Signal(QtGui.QFont)
```

🚫  🎲

**Comment by klauer** [comment]
@hhslepicka can you fix the flake8 errors, though? I think this will be ready then.

**Comment by hhslepicka** [comment]
@klauer sorry about that... Issues addressed now.

**Pull request merged**